File: java/array_search/LinearSearchProduction.java
import java.util.List;
import java.util.Objects;
import java.util.function.Predicate;

public final class LinearSearchProduction {
    public record SearchResult(boolean found, int index, int inspected) {
    }

    private LinearSearchProduction() {
    }

    public static <T> SearchResult findFirst(List<T> values, T target) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }

        return findFirst(values, target, 0, values.size());
    }

    public static <T> SearchResult findFirst(List<T> values, T target, int start, int stop) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }

        validateRange(values.size(), start, stop);
        int inspected = 0;

        for (int index = start; index < stop; index++) {
            inspected += 1;

            if (Objects.equals(values.get(index), target)) {
                return new SearchResult(true, index, inspected);
            }
        }

        return new SearchResult(false, -1, inspected);
    }

    public static <T> SearchResult findFirstWhere(List<T> values, Predicate<? super T> predicate) {
        if (values == null || predicate == null) {
            throw new IllegalArgumentException("values and predicate must not be null");
        }

        return findFirstWhere(values, predicate, 0, values.size());
    }

    public static <T> SearchResult findFirstWhere(List<T> values, Predicate<? super T> predicate, int start, int stop) {
        if (values == null || predicate == null) {
            throw new IllegalArgumentException("values and predicate must not be null");
        }

        validateRange(values.size(), start, stop);
        int inspected = 0;

        for (int index = start; index < stop; index++) {
            inspected += 1;

            if (predicate.test(values.get(index))) {
                return new SearchResult(true, index, inspected);
            }
        }

        return new SearchResult(false, -1, inspected);
    }

    private static void validateRange(int length, int start, int stop) {
        if (start < 0 || start > length || stop < start || stop > length) {
            throw new IllegalArgumentException("range must satisfy 0 <= start <= stop <= values.size()");
        }
    }
}

File: python/array_search/linear_search_production.py
from dataclasses import dataclass 


@dataclass(frozen=True)
class LinearSearchResult:
    found: bool 
    index: int 
    inspected: int


def find_first(values, target, *, key=None, start=0, stop=None):
    key_fn = key or (lambda value: value)
    length = len(values)
    end = length if stop is None else min(stop, length)

    if start < 0 or start > length:
        raise ValueError("start must be between 0 and len(values)")

    if end < start: 
        raise ValueError("stop must be greater than or equal to start")

    inspected = 0 

    for index in range(start, end):
        inspected += 1

        if key_fn(values[index]) == target:
            return LinearSearchResult(True, index, inspected)

    return LinearSearchResult(False, -1, inspected)


def find_first_where(values, predicate, *, start=0, stop=None):
    length = len(values)
    end = length if stop is None else min(stop, length)

    if start < 0 or start > length:
        raise ValueError("start must be between 0 and len(values)")

    if end < start: 
        raise ValueError("stop must be greater than or equal to start")

    inspected = 0 

    for index in range(start, end):
        inspected += 1 

        if predicate(values[index]):
            return LinearSearchResult(True, index, inspected)

    return LinearSearchResult(False, -1, inspected)
